lib/puppet/pops/types/type_factory.rb

```diff
@@ -181,13 +181,7 @@ def self.any
   # @api public
   #
   def self.regexp(pattern = nil)
-    if pattern
-      t = PRegexpType.new(pattern.is_a?(Regexp) ? pattern.inspect[1..-2] : pattern)
-      t.regexp unless pattern.nil? # compile pattern to catch errors
-      t
-    else
-      PRegexpType::DEFAULT
-    end
+    pattern ?  PRegexpType.new(pattern) : PRegexpType::DEFAULT
```

Comment:
The above change ensures that a `Regexp` instance can pass through without intervention. This is OK because the `PRegexpType` constructor will accept either a `String` or a `Regexp`. The actual string representation is not dictated by the instance itself, but by which method that is chosen to represent it (#inspect, #source, #to_s). The `Regexp` instance itself keeps the `source` and `options` separate.

Prior to this change, the `Regexp` was converted to a `String` using #inspect which always produces the `/pat/x` form. The subsequent substring expression then converted the string to "pat/". Although still a valid expression, it no longer represents the original expression. I.e. the regexp is now corrupted.

By passing the `Regexp` directly to the `PRegexpType` constructor, the responsibility of producing the string representation is moved. The `PRegexType` infers the string by either using the `#source` method (which doesn't include the options) or by using the `#to_s` method which will include the options using the `/(?x:pat)/` syntax. The reason for not always using the `#to_s` method is that it will _always_ include the options, even when no options have been set (as `-xmi`).